File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cs
public class Solution {
    public int LengthOfLongestSubstring(string s) {
        if(s == null | s.Length == 0)
            return 0;
        Dictionary<char,int> map = new Dictionary<char,int>();
        
        int maxlen = 0, clen = 0;
        int start = 0;
        for(int i = 0; i < s.Length; i++){
            if(!map.ContainsKey(s[i])){
                map.Add(s[i], i);
                clen = i-start+1;
                maxlen = Math.Max(maxlen, clen);
            }
            else{
                int lastindex = map[s[i]];
                
                while(start <= lastindex){
                    if(map.ContainsKey(s[start])){
                        map.Remove(s[start]);
                    }
                    
                    start++;
                }
                
                map.Add(s[i],i);
                
                clen = i-start+1;
                maxlen = Math.Max(maxlen, clen);
            }
        }
        
        return maxlen;
    }
}

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int len = s.length();
        int max = 0;
        
        Map<Character,Integer> map = new HashMap<>();
        //aabccbb
        int low = 0, hi = 0;
        while(hi < len){
            if(!map.containsKey(s.charAt(hi))){
                map.put(s.charAt(hi), hi);
            }
            else{
                int lastIndex = map.get(s.charAt(hi));
                System.out.println(lastIndex);
                while(low <= lastIndex){
                    map.remove(s.charAt(low));
                    low++;
                }
                
                low = lastIndex+1;
                map.put(s.charAt(hi), hi);
            }
            
            max = Math.max(max, hi-low+1);
            hi++;
        }
        
        return max;
    }
}